implicit/als.py

```diff
@@ -105,12 +105,36 @@ def recommend(self, userid, user_items, N=10, filter_items=None, recalculate_use
     def _user_factor(self, userid, user_items, recalculate_user=False):
         if not recalculate_user:
             return self.user_factors[userid]
-        Y = self.item_factors
-        if self._YtY is None:
-            self._YtY = Y.T.dot(Y)
-        return user_factor(Y, self._YtY, user_items, userid,
+        return user_factor(self.item_factors, self.YtY, user_items, userid,
                            self.regularization, self.factors)
 
+    def explain(self, userid, user_items, itemid, user_weights=None, N=10):
+        """ Returns the predicted rating for an user x item pair,
+            the explanation (the contribution from the top N items the user liked),
+            and a user latent factor weight that can be cached if you want to
+            get more than one explanation for the same user.
+        """
+        if user_weights is None:
+            A, _ = user_linear_equation(self.item_factors, self.YtY,
+                                        user_items, userid,
+                                        self.regularization, self.factors)
+            user_weights = np.linalg.inv(A)
+
+        seed_item = self.item_factors[itemid]
+        weighted_item = seed_item.T.dot(user_weights)
+
+        contributions = []
+        total_score = 0.0
+        for i, confidence in nonzeros(user_items, userid):
+            factor = self.item_factors[i]
+            similarity = weighted_item.dot(factor)
+            score = similarity * confidence
+            total_score += score
+            contributions.append((i, score))
+
+        top_contributions = sorted(contributions, key=lambda x: -x[1])[:N]
```

Comment:
Instead of using `nlargest`, I ended up using `pushpop` so that we never have a heap larger than `N`